### PR TITLE
FilterPopover

### PR DESCRIPTION
**Pull request by WoH**
- FAB button
- Basic Popover (with inline Template -> https://github.com/driftyco/ionic/issues/6797) with dual knob Slider
- Introduce a filter object in MapPage
- Sync time filter between MapPage and FilterPopover (debounced)
- MapPage testing
- FilterPopover testing

### CODE REVIEW

**Comment by johartl** [comment]
I haven't really looked into it yet but I think we need to discuss how we structure our components properly. 
I saw that you put the `filterpopover.ts` into `/app/pages/map/filterpopover`. In #40 @AlexanderLill placed his component into the root folder `/app` and named the file `app-menu.component.ts`. So for the sake of consistency I think we should come to common understanding of where each file belongs to.

I would suggest the following structure

```
/app
  /components
    /filter-popover
      filter-popover.component.ts
      filter-popover.component.html
      filter-popover.component.scss
  /pages
    /map
      map.page.ts
      map.page.html
      map.page.scss
  /services
    api.service.ts
  /themes (renamed from theme)
    app.core.scss
```

Here are a few arguments for the proposed structure
- Why the `.{component|page|service}.ts` suffixes? Because it's strongly recommended by the Angular 2 style guide. You said that the Ionic people tend to treat things different here but in this case I would like to give a higher precedence to Angular. After all, Ionic is just a framework that makes use of Angular so I think they should adhere to their style guide as well. Suffixes also prevent name clashes, for instance we have a `MapPage` but I also wanted to create a `MapComponent` that wraps the PokeMap from the other team.
- Why are the components not placed in the `/pages/XYZ` folder? Components are all about reusability. I understand that you put the FAB component inside the `/page/map` folder because it only appears on the map page. But this isn't necessary the case for all components. For instance, the navbar #26  will appear in all pages so in which folder should we place it? Putting them all together in `/components` means that we don't have to bother about such problems.

What do you think? @WoH @AlexanderLill

**Comment by WoH** [comment]
Only other reasonable idea I can think of:

```
app
 |-- components
 |    |-- pages
 |    |    |-- map
 |    |    |-- wiki
 |    |    +-- about
 |    +-- popovers
 |         +-- filter
 |-- services
 +-- directives
```

**Comment by WoH** [comment]
My reasoning was to put it in a subfolder of the map page bc. we will only need it as part of the map page.
But we could theoretically just look at it as a reusable Popover

**Comment by johartl** [comment]
I think the `/app/components/pages/map` path is a bit deep. And what exactly makes a component a `popover`? I feel like this will lead to endless discussions about where to put a specific component.
Ideally when you create a new component there is no doubt about where to put it. That's what I like about placing everything inside `/app/components`.

**Comment by WoH** [comment]
Other way around, a Popover is a Component, that's why it's a subfolder, just like Pages are Components

**Comment by WoH** [comment]
Ionic suggests (from my understanding, I haven't seen a lot of Popovers):

```
app
 |-- pages
 |    |-- map
 |    |-- wiki
 |    +-- about
 +-- providers
 |    |-- api-service
 |    +-- filter-popover
```

**Comment by johartl** [comment]
Yes I understand. But then every component must belong to a "subcategory", so it's either a Page, a Popover, ...? I just think using popover as a subfolder is really arbitrary. As you said, there probably won't be a lot of Popover components out there ;)

Thanks for posting the Ionic standard. To me `providers` are only those classes that are declared under the providers key in the Component decorator:

``` (javascript)
@Component({
    providers: [/* list of providers */]
})
export class App { ... }
```

The `filter-popover` component wouldn't be listed as a provider though. That's why I think this name is very misleading.

**Comment by WoH** [comment]
The Services should 100% end up there (app/providers). But I can't find any suggestions for the popover.
I know Modals are usually just a subfolder to app/pages.

**Comment by johartl** [comment]
Oh we also have some plain JS classes that serve as models (`Pokemon`, `Sighting`, `PokemonIcon` so far). I would put these into `/app/models`.
Is there anything you don't like about putting all (non-page) components into `/app/components`?

**Comment by WoH** [comment]
Just a bit of a sour taste if Pages, that are theoretically components are not in that subfolder. But for the sake of getting a guideline in place, which would be a neccessary thing, let's do it

**Comment by WoH** [comment]
I'll refactor here to meet the guidelines for this one before we merge.
Still would appreciate feedback on the implementation 😉

**Comment by johartl** [comment]
```
/app
  /components
    /filter-popover
      filter-popover.component.ts (class FilterPopoverComponent)
      filter-popover.component.html
      filter-popover.component.scss
  /models
    pokemon.ts (class Pokemon)
  /pages
    /map
      map.page.ts (class MapPage)
      map.page.html
      map.page.scss
  /services
    api.service.ts (class AppService)
  /themes (renamed from theme)
    app.core.scss
```

Is that what we agreed on?

I will give you feedback on the FAB later tonight.

**Comment by johartl** [comment]
Code looks really fine to me and very nice test cases :+1: Just two points
- I'm not sure how often prediction will occur in practice but shouldn't we be using minutes rather than days? Or possibly 10, 30 or 60 minute steps? I'm probably more interested in the Pokemon that will spawn in the next hour, not in 2-3 weeks.
- I would also update the header text `Time (days)` in the popover to something more meaningful like `Show predictions for time range (days)`.

**Comment by WoH** [comment]
I agree, I didn't update these yet. The conversation we had happened after my initial draft.
Thanks for reminding me!

File: ionic2/app/components/filter-popover/filter-popover.component.ts [hunk 1]
```diff
@@ -4,15 +4,15 @@ import { Events, ViewController } from 'ionic-angular';
 @Component({
   template: `
   <ion-list style="margin-bottom: 0">
-    <ion-list-header>Time (days):</ion-list-header>
+    <ion-list-header>Limit prediciton time (minutes):</ion-list-header>
```
Comment (johartl) [inline ionic2/app/components/filter-popover/filter-popover.component.ts]: You have a typo here

**Comment by johartl** [comment]
After you fixed the typo it's good to merge.

**Comment by johartl** [comment]
You can go ahead and merge it!

**Pull request merged**